larq/metrics.py

```diff
@@ -0,0 +1,99 @@
+import tensorflow as tf
+from larq import utils
+import numpy as np
+
+
+class MeanChangedValues(tf.keras.metrics.Mean):
+    """Computes the mean ration of changed values in a given tensor.
+
+    !!! example
+        ```python
+        m = metrics.MeanChangedValues()
+        m.update_state(1)
+        m.update_state(2)
+        m.update_state(2)
+        print('Final result: ', m.result().numpy())  # Final result: 0.5
+        ```
+
+    # Arguments
+    values_shape: Shape of the tensor for which to track changes.
+    values_dtype: Data type of the tensor for which to track changes.
+    name: Name of the metric.
+    dtype: Data type of the moving mean.
+    """
+
+    def __init__(
+        self,
+        values_shape=(),
+        values_dtype="int8",
+        name="mean_changed_values",
+        dtype=None,
+    ):
+        super().__init__(name=name, dtype=dtype)
+        self.values_dtype = tf.as_dtype(values_dtype)
+        self.values_shape = tf.TensorShape(values_shape).as_list()
+        self._previous_values = self.add_weight(
+            "previous_values",
+            shape=values_shape,
+            dtype=self.values_dtype,
+            initializer=tf.keras.initializers.zeros,
+        )
+        self._size = np.prod(self.values_shape)
+        self._built = False
+
+    def update_state(self, values, sample_weight=None):
+        values = tf.cast(values, self.values_dtype)
+        if not self._built:
+            self._built = True
+            return self._previous_values.assign(values)
+        changed_values = tf.math.count_nonzero(tf.equal(self._previous_values, values))
+        metric_update_op = super(MeanChangedValues, self).update_state(
+            1 - (tf.cast(changed_values, tf.float32) / self._size)
+        )
+        with tf.control_dependencies([metric_update_op]):
+            return self._previous_values.assign(values)
+
+    def reset_states(self):
+        tf.keras.backend.batch_set_value(
+            [(v, 0) for v in self.variables if v != self._previous_values]
+        )
+
+    def get_config(self):
+        return {
+            **super().get_config(),
+            "values_shape": self.values_shape,
+            "values_dtype": self.values_dtype.name,
+        }
+
+    def add_weight(
```

Comment:
All of this code is really annoying, but the alternative would be not to cast the previous weights to a smaller dtype which might or might not increase memory consumption.